libs/elasticsearch-nio/src/main/java/org/elasticsearch/nio/EventHandler.java

```diff
@@ -20,15 +20,171 @@
 
-    protected EventHandler(Consumer<Exception> exceptionHandler) {
+    public EventHandler(Consumer<Exception> exceptionHandler) {
+        this(exceptionHandler, null);
+    }
+
+    public EventHandler(Consumer<Exception> exceptionHandler, Supplier<NioSelector> selectorSupplier) {
         this.exceptionHandler = exceptionHandler;
+        this.selectorSupplier = selectorSupplier;
+    }
+
+    /**
+     * This method is called when a server channel signals it is ready to accept a connection. All of the
+     * accept logic should occur in this call.
+     *
+     * @param context that can accept a connection
+     */
+    protected void acceptChannel(ServerChannelContext context) throws IOException {
+        if (selectorSupplier != null) {
+            context.acceptChannels(selectorSupplier);
+        } else {
+            context.acceptChannels(context::getSelector);
+        }
+    }
+
+    /**
+     * This method is called when an attempt to accept a connection throws an exception.
+     *
+     * @param context that accepting a connection
+     * @param exception that occurred
+     */
+    protected void acceptException(ServerChannelContext context, Exception exception) {
+        context.handleException(exception);
+    }
+
+    /**
+     * This method is called when a NioChannel is being registered with the selector. It should
+     * only be called once per channel.
+     *
+     * @param context that was registered
+     */
+    protected void handleRegistration(ChannelContext<?> context) throws IOException {
+        context.register();
+        SelectionKey selectionKey = context.getSelectionKey();
+        selectionKey.attach(context);
+        if (context instanceof SocketChannelContext) {
+            if (((SocketChannelContext) context).readyForFlush()) {
+                SelectionKeyUtils.setConnectReadAndWriteInterested(context.getSelectionKey());
+            } else {
+                SelectionKeyUtils.setConnectAndReadInterested(context.getSelectionKey());
+            }
+        } else {
+            assert context instanceof ServerChannelContext : "If not SocketChannelContext the context must be a ServerChannelContext";
+            SelectionKeyUtils.setAcceptInterested(context.getSelectionKey());
+        }
+    }
+
+    /**
+     * This method is called when an attempt to register a channel throws an exception.
+     *
+     * @param context that was registered
+     * @param exception that occurred
+     */
+    protected void registrationException(ChannelContext<?> context, Exception exception) {
+        context.handleException(exception);
+    }
+
+    /**
+     * This method is called when a NioSocketChannel has just been accepted or if it has receive an
+     * OP_CONNECT event.
+     *
+     * @param context that was registered
+     */
+    protected void handleConnect(SocketChannelContext context) throws IOException {
+        if (context.connect()) {
```

Comment:
why would this be called if it's not a connect context? should we handle this on the caller side?

Reply:
I'm not sure I follow this comment. `context.connect()` implements the actual connection logic. If it returns true, that means the connection is complete.